R/bundlePackage.R

```diff
@@ -103,75 +102,72 @@ findPackageRepoAndSource <- function(records, repos) {
-  tmp <- lapply(seq_len(nrow(records)), function(i) {
+standardizePackageRepoAndSource <- function(record, availablePackages) {
+  pkg <- record$Package
+  source <- record$Source
 
-    pkg <- records[i, "Package"]
-    source <- records[i, "Source"]
+  # source types are defined by packrat:
+  # https://github.com/rstudio/packrat/blob/v0.9.0/R/pkg.R#L328
+  if (source %in% c("github", "gitlab", "bitbucket")) {
+    # SCM information is recorded elsewhere
     repository <- NA
-    # capture Bioconcutor repository
-    if (identical(source, "Bioconductor")) {
-      if (pkg %in% biocPackages) {
-        repository <- biocPackages[pkg, "Repository"]
-      }
-    } else if (isSCMSource(source)) {
-      # leave source+SCM packages alone.
-    } else if (pkg %in% rownames(repo.packages)) {
-      # capture CRAN-like repository
-
-      # Find this package in the set of available packages then use its
-      # contrib.url to map back to the configured repositories.
-      package.contrib <- repo.packages[pkg, "Repository"]
-      package.repo.index <- vapply(repo.lookup$contrib.url,
-                                   function(url) grepl(url, package.contrib, fixed = TRUE), logical(1))
-      package.repo <- repo.lookup[package.repo.index, ][1, ]
-      # If the incoming package comes from CRAN, keep the CRAN name in place
-      # even if that means using a different name than the repos list.
-      #
-      # The "cran" source is a well-known location for shinyapps.io.
-      #
-      # shinyapps.io isn't going to use the manifest-provided CRAN URL,
-      # but other consumers (Connect) will.
-      if (tolower(source) != "cran") {
-        source <- package.repo$name
-      }
-      repository <- package.repo$url
+  } else if (source == "source") {
+    # can't install source packages elsewhere
+    source <- NA
+    repository <- NA
+  } else if (source == "CustomCRANLikeRepository") {
+    # Packrat guessed that this package was installed from CRAN, but
+    # we need to check that it's not actually a development version
+    if (is_dev_version(record, availablePackages)) {
+      source <- NA
+      repository <- NA
+    } else {
+      repository <- findRepo(availablePackages, pkg)
     }
-    # validatePackageSource will emit a warning for packages with NA repository.
-    data.frame(Source = source, Repository = repository, stringsAsFactors = FALSE)
-  })
-  do.call("rbind", tmp)
+  } else {
+    # Installed from CRAN, BioC, or another repository
+    # NB: shinyapps will ignore repository if source is CRAN or BioC
+    repository <- findRepo(availablePackages, pkg)
+  }
+  list(Source = source, Repository = repository)
 }
 
+findRepo <- function(availablePackages, pkg) {
+  if (pkg %in% rownames(availablePackages)) {
+    repo <- availablePackages[pkg, "Repository"]
+    # Strip /src/contrib from repository URL recorded in package record to
+    # get to repository url
+    gsub("/src/contrib$", "", repo)
```

Comment:
It looks like `/src/contrib` is always the path appended by `contrib.url` for "source". I wasn't sure previously, which is why the old version maps from `url` to `contrib.url(url."source")`.

https://github.com/wch/r-source/blob/trunk/src/library/utils/R/packages.R#L832-L866